common/src/main/java/org/conscrypt/ClientSessionContext.java

```diff
@@ -140,12 +169,10 @@ void onBeforeAddSession(NativeSslSession session) {
         }
 
         HostAndPort key = new HostAndPort(host, port);
-        synchronized (sessionsByHostAndPort) {
-            sessionsByHostAndPort.put(key, session);
-        }
+        putSession(key, session);
 
         // TODO: Do this in a background thread.
-        if (persistentCache != null) {
+        if (persistentCache != null && !session.isSingleUse()) {
```

Comment:
There's a subtlety here: if the new session is single-use, we additionally clobber the existing multiple-use session in Chrome (and we'll never keep more than one multiple-use session around).

Consider if a server upgrades from TLS 1.2 to TLS 1.3. Previously, I had a trackable TLS 1.2 session, with the privacy leak inherent in the protocol. Now it upgrades to TLS 1.3 with single-use tickets. It mints me single-use tickets that I now add to the session cache. The next connections will use those and I'm happy.

But if I run out of single-use tickets and I've kept the 1.2 session around, I'll fall back to offering the 1.2 session rather than nothing, restoring the privacy leak. That 1.2 session is not even useful; if the server's updated to TLS 1.3, it will most likely[*] ignore the TLS 1.2 session anyway.

(Granted, ideally your TLS 1.2 sessions are short-lived anyway, in which case this isn't a big deal as it doesn't last more than an hour or so.)

[*] Assuming it hasn't rolled back TLS 1.3 for whatever reason.

Reply:
Good point, updated so it doesn't mix single- and multi-uses sessions for the same endpoint.